WT-14931 Enable the preserve prepared config in test/format and test/checkpoint

tdlinh2712 opened the pull request:
Enable the preserve prepared config in test format and test checkpoint. Set prepared id and rollback timestamps for prepared transactions.

tdlinh2712 committed:
WT-14931 enable prepare operation in test/format

tdlinh2712 committed:
add preserve_prepared config to test checkpoint

quchenhao commented:
@tdlinh2712 I think we also need to change test format code. Adding the preseve_prepared config to test format config. There is also some i code I put inside test format to not run prepared when precise checkpoint is enabled. We need to remove that. In test format, we need to ensure we don't run preserve prepared when precise checkpoint is not enabled.

quchenhao commented on test/evergreen.yml:
```diff
@@ -3804,7 +3804,7 @@ tasks:
       - func: "run data validation stress test checkpoint"
         vars:
-          run_test_checkpoint_args: -x -C cache_size=1000MB,checkpoint=(precise=true)
+          run_test_checkpoint_args: -x -C cache_size=1000MB,checkpoint=(precise=true),preserve_prepared=true
```
We should also pass in the -p flag to enable prepared.

quchenhao committed:
Merge branch 'develop' into wt-14931

quchenhao committed:
Enable prepare for test checkpoint

quchenhao committed:
Add preserve prepared to test format

quchenhao committed:
Merge branch 'develop' into wt-14931

quchenhao committed:
No need to do rts is no stable ts and precise checkpoint is enabled

quchenhao committed:
Fix loop

quchenhao committed:
Fix test checkpoint

quchenhao committed:
Fix compile

quchenhao committed:
Fix compile

quchenhao committed:
Fix test format

quchenhao committed:
Fix test format

quchenhao committed:
Refactor timestamp parsing

quchenhao committed:
Fix compile

quchenhao committed:
Support prepared id and rollback ts in test format

quchenhao committed:
Fix test_timestamp27

quchenhao committed:
Separate the config

quchenhao committed:
Fix s_all

quchenhao committed:
Fix compile

quchenhao committed:
Fix compile

quchenhao committed:
Fix compile

quchenhao committed:
Fix ex_all

quchenhao committed:
Fix ex_all

quchenhao committed:
Fix python test failures

quchenhao committed:
Fix test_timestamp01

quchenhao committed:
Merge branch 'develop' into wt-14931

quchenhao committed:
Fix an assert

quchenhao committed:
Fix compile

quchenhao committed:
Fix prepared id increment

quchenhao committed:
Init prepared id

quchenhao committed:
Merge branch 'develop' into wt-14931

quchenhao committed:
Fix prepare timestamp conflict with stable timestamp

quchenhao committed:
WT-15238 Fix prepared commit and rollback race with reconciliation

quchenhao committed:
Fix comment

quchenhao committed:
Remove an assert that is no longer valid

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Remove a wrong assert

quchenhao committed:
Fix format tests

quchenhao committed:
Add back an assert

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix some issues

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix discard the global visible tombstone

quchenhao committed:
Fix discard the global visible tombstone

quchenhao committed:
Fix issues

quchenhao committed:
Fix issues

quchenhao committed:
Add a test

quchenhao committed:
Fix assert failure

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix a potential race

quchenhao committed:
Fix other races

quchenhao committed:
Handle obsolete check

quchenhao committed:
Fix another race

quchenhao committed:
Fix assert

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix another race

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix test skip

quchenhao committed:
Fix import

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Improve checking

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix dangling prepared updates

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix crash

quchenhao committed:
Merge branch 'wt-15238' into wt-14931

quchenhao committed:
Fix column store logic

quchenhao committed:
Merge branch 'develop' into wt-14931

quchenhao committed:
Fix merge error

tdlinh2712 committed:
WT-15342 Re-enable RTS for precise checkpoint to unblock test format (#12316)

Re-enable RTS for precise checkpoint to unblock test format. Disable
this setting when claim prepared is implemented in test format

quchenhao committed:
Merge branch 'develop' into wt-14931

quchenhao committed:
Increase cache for precise checkpoint and disable prepare for disagg

Siddhartha8899 commented on test/format/CONFIG.disagg:
```diff
@@ -18,8 +18,6 @@ ops.compaction=0
 ops.throttle=0
 ops.truncate=0
-# FIXME-WT-14931 should allow prepared operations for layered tables.
-ops.prepare=0
```
We should enable prepare ops for disagg in test/format, this creates failures in Disagg mode. We should link [WT-15051](https://jira.mongodb.org/browse/WT-15051) and fix this first. As per Hari prepare tranactions are low prior for disagg, we can push this for some time.

quchenhao replied:
oh, ok. I have it disabled somewhere else. I'll revert this.

quchenhao committed:
Disable prepare for disagg the old way

quchenhao committed:
Fix cache size

Siddhartha8899 commented on test/evergreen.yml:
```diff
@@ -3881,7 +3881,7 @@ tasks:
       - func: "run data validation stress test checkpoint"
         vars:
-          run_test_checkpoint_args: -x -s 7 -C cache_size=1000MB,precise_checkpoint=true
+          run_test_checkpoint_args: -x -s 7 -C cache_size=1000MB,precise_checkpoint=true,preserve_prepared=true
```
is there value in having seperate tasks for prepared oriented tests? It seems valuable to have non-prepared tests.

Note: test/checkpoint tests are scattered around a bit, there are a set of tests defined test/checkpoint/CMakeLists.txt as welll

quchenhao replied:
I would say prepare test tests both prepare and non-prepare code path. I don't think we need to add another test.

Siddhartha8899 replied:
Sure, Did you want this support to be present across all the test/checkpoint tests? or dleiberately just the data validation test/checkpoint tasks?

There are more tests in test/checkpoint/CMakeLists.txt

quchenhao replied:
just data validation test. I need the test to verify after restart.

Siddhartha8899 commented on test/format/t.c:
```diff
@@ -340,13 +340,15 @@ main(int argc, char *argv[])
         testutil_assert(ret == 0 || ret == WT_NOTFOUND);
         ret = timestamp_query("get=stable_timestamp", &g.stable_timestamp);
         testutil_assert(ret == 0 || ret == WT_NOTFOUND);
+        g.prepared_id = 1;
```
can we find a place for this in function `config_run`.  In config_transaction?

quchenhao replied:
what do you mean? move this to config_transaction. ok

quchenhao replied:
done

quchenhao committed:
Move the init of prepared id to config_transaction

quchenhao committed:
Fix compile

Siddhartha8899 commented on test/format/format.h:
```diff
@@ -413,6 +414,7 @@ typedef struct {
     uint64_t read_ts;      /* read timestamp */
     uint64_t replay_ts;    /* allocated timestamp for predictable replay */
     uint64_t stable_ts;    /* stable timestamp */
+    uint64_t prepared_id;  /* prepared id */
```
do we use this?

quchenhao replied:
yes

Siddhartha8899 commented on test/checkpoint/workers.c:
```diff
@@ -468,8 +470,10 @@ real_worker(THREAD_DATA *td)
+                            prepared_id = __wt_atomic_addv64(&g.prepared_id, 1);
+                            testutil_snprintf(buf, sizeof(buf),
+                              "prepare_timestamp=%" PRIx64 ",prepared_id=%" PRIx64, base_ts,
+                              prepared_id);
```
prepare id stays constant?

quchenhao replied:
prepared_id = __wt_atomic_addv64(&g.prepared_id, 1); it is incremented for each transaction

Siddhartha8899 commented on test/format/wts.c:
```diff
@@ -784,13 +789,13 @@ wts_open(const char *home, WT_CONNECTION **connp, bool verify_metadata)
+        if (GV(PRECISE_CHECKPOINT))
             CONFIG_APPEND(p, ",precise_checkpoint=true");
 
+        if (GV(PRESERVE_PREPARED))
```
this is also present in create `wts_open`. Also, should we check if ops.prepared is enabled and only then we check is preserve_prepared is true?

quchenhao replied:
there will be no issue if prepared is not enabled. It is just an no op. I have to do this in both here and wts_open to ensure we pass the correct config. wts_open is called for reopen and we need to reopen with the same config.

Siddhartha8899 replied:
I believe its important for readability. Even if its a no-op. What do you think?

Siddhartha8899 replied:
> I have to do this in both here and wts_open to ensure we pass the correct config. wts_open is called for reopen and we need to reopen with the same config.

wts_open seems to be called both here and reopen? Sorry I have misunderstood something

quchenhao replied:
I can add a comment?

Siddhartha8899 commented on test/format/wts.c:
```diff
@@ -680,17 +686,16 @@ create_object(TABLE *table, void *arg)
  */
 static void
-disagg_conn_init(WT_CONNECTION *conn)
+precise_checkpoint_init(WT_CONNECTION *conn)
```
just confirming, if disagg is enabled without precise checkpoint this could be a problem?

Siddhartha8899 commented on test/format/format_config_def.c:
```diff
@@ -298,7 +298,9 @@ CONFIG configuration_list[] = {{"assert.read_timestamp", "assert read_timestamp"
-  {"precise_checkpoint", "Precise checkpoint", C_BOOL, 50, 0, 0, V_GLOBAL_CHECKPOINT_PRECISE},
+  {"precise_checkpoint", "Precise checkpoint", C_BOOL, 50, 0, 0, V_GLOBAL_PRECISE_CHECKPOINT},
+
+  {"preserve_prepared", "Preserve prepared", C_BOOL, 50, 0, 0, V_GLOBAL_PRESERVE_PREPARED},
```
Does the 50 here represent the probability, so every other run enable this?

quchenhao replied:
yes, I think so

quchenhao committed:
Address review comment
